src/finn/transformation/fpgadataflow/templates.py

```diff
@@ -126,120 +126,113 @@ def __init__(
-            ibuf_folded, idt, reverse_endian=True, reverse_inner=True
+            ibuf_folded, self.idt, reverse_endian=True, reverse_inner=True
         )
         return ibuf_packed
 
     def unpack_output(self, obuf_packed):
         N = self.N
-        # output FINN DataType
-        odt = $OUTPUT_FINN_DATATYPE$
-        oshape_folded = $OUTPUT_SHAPE_FOLDED$
         # unpack the packed output buffer from accelerator
         obuf_folded = packed_bytearray_to_finnpy(
-            obuf_packed, odt, oshape_folded, reverse_endian=True, reverse_inner=True
+            obuf_packed, self.odt, self.oshape_folded, reverse_endian=True, reverse_inner=True
         )
         return obuf_folded
 
-    def save_output(self, obuf_folded):
+    def save_output(self, obuf_folded, outputfile):
         N = self.N
         # convert to normal reshape and save
-        oshape_normal = $OUTPUT_SHAPE_NORMAL$
-        obuf_normal = obuf_folded.reshape(oshape_normal)
-        np.save(self.outputfile, obuf_normal)
-
-    def allocate_pynqbuffer(self, shape, data=None):
-        buf_device = allocate(shape=shape, dtype=np.uint8)
+        obuf_normal = obuf_folded.reshape(self.oshape_normal)
+        np.save(outputfile, obuf_normal)
 
+    def allocate_pynqbuffers(self, data=None):
+        ibuf_packed_device = allocate(shape=self.ishape_packed, dtype=np.uint8)
         # if necessary copy the packed data into the PYNQ buffer
         # TODO optimization: pack directly into the PYNQ buffer?
         if data is not None:
-            np.copyto(buf_device, data)
-
-        return buf_device
+            np.copyto(ibuf_packed_device, data)
+        obuf_packed_device = allocate(shape=self.oshape_packed, dtype=np.uint8)
 
+        return [ibuf_packed_device, obuf_packed_device]
 
-    def run_nw(self):
-        exec_mode = self.exec_mode
-        if exec_mode == "remote_pynq":
-            ibuf_normal = self.load_input()
-            ibuf_packed = self.pack_input(ibuf_normal)
-        elif exec_mode != "throughput_test":
-            raise Exception("Exec mode has to be set to remote_pynq or throughput_test")
-
+    def setup_TLastMarker(self):
+        N = self.N
         # set up TLastMarker with correct num. samples
         self.ctrl_regs.write(self.REG_OFFSET_NUM_ITERS, N*self.itersPerSample)
 
-        # allocate a PYNQ buffer for the packed input buffer
-        if exec_mode == "remote_pynq":
-            ibuf_packed_device = self.allocate_pynqbuffer(self.ishape_packed, ibuf_packed)
-        else:
-            ibuf_packed_device = self.allocate_pynqbuffer(self.ishape_packed)
-
-        # allocate a PYNQ buffer for the returned packed output buffer
-        obuf_packed = self.allocate_pynqbuffer(self.oshape_packed)
-
-        if exec_mode == "throughput_test":
-            # measure runtime of network
-            start = time.time()
-            res={}
 
+    def run_nw(self, ibuf_packed_device, obuf_packed_device):
```

Comment:
rename this fxn execute
remove ibuf/obuf arguments, they will be class members (since we assume buffers that "belong" to the driver instance) and there is no need to specify them